### ISSUE TITLE
[REVIEW]: Makie.jl: Flexible high-performance data visualization for Julia

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3349]
**Submitting author:** @jkrumbiegel (<a href="http://orcid.org/0000-0002-8409-5590">Julius Krumbiegel</a>)
**Repository:** <a href="https://github.com/JuliaPlots/Makie.jl" target ="_blank">https://github.com/JuliaPlots/Makie.jl</a>
**Version:** v.0.15.0
**Editor:** @Kevin-Mattheus-Moerman
**Reviewer:** @fverdugo, @gaelforget
**Archive:** <a href="https://doi.org/10.5281/zenodo.5106448" target="_blank">10.5281/zenodo.5106448</a>

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/4fec1fdaaa4b7d3e604d761e5258c275/status.svg)](https://joss.theoj.org/papers/4fec1fdaaa4b7d3e604d761e5258c275)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/4fec1fdaaa4b7d3e604d761e5258c275"><img src="https://joss.theoj.org/papers/4fec1fdaaa4b7d3e604d761e5258c275/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/4fec1fdaaa4b7d3e604d761e5258c275/status.svg)](https://joss.theoj.org/papers/4fec1fdaaa4b7d3e604d761e5258c275)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/JuliaPlots/AbstractPlotting.jl" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@fverdugo & @gaelforget, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @Kevin-Mattheus-Moerman know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/JuliaPlots/AbstractPlotting.jl">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@jkrumbiegel) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/JuliaPlots/AbstractPlotting.jl">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@jkrumbiegel) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @fverdugo, @gaelforget it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.37 s (578.2 files/s, 96509.1 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Julia                          128           4367           1582          22536
Markdown                        73           2054              0           5122
YAML                             9              6             10            219
TOML                             6              6              0            151
-------------------------------------------------------------------------------
SUM:                           216           6433           1592          28028
-------------------------------------------------------------------------------


Statistical information for the repository '268a670e3edea2abe216abb7' was
gathered on 2021/06/09.
No commited files with the specified extensions were found.
```

**Comment by whedon** [comment]
PDF failed to compile for issue #3349 with the following error: 
```
 Can't find any papers to compile :-(

```

**Comment by Kevin-Mattheus-Moerman** [comment]
@jkrumbiegel @gaelforget @fverdugo this is where the review will take place.

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon generate pdf from branch joss-paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss-paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03349/joss.03349/10.21105.joss.03349.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03349/joss.03349/10.21105.joss.03349.pdf) :page_facing_up: :point_left:

**Comment by jkrumbiegel** [comment]
@Kevin-Mattheus-Moerman I just wanted to let you know that we have recently consolidated our codebase into the https://github.com/JuliaPlots/Makie.jl/ repository, whereas before, most code lived in `AbstractPlotting.jl`. The paper lies in the AbstractPlotting repository, still, and I'm not sure if it's worth the trouble of moving and relinking it. But if you think that would be better long term, let me know. It wasn't foreseeable for me at the time I submitted the paper that we would start this major internal refactor so soon.

**Comment by arfon** [comment]
@jkrumbiegel – please move your paper into the new repository (https://github.com/JuliaPlots/Makie.jl)

**Comment by jkrumbiegel** [comment]
@arfon All right, I have moved the paper to https://github.com/JuliaPlots/Makie.jl/tree/joss-paper

**Comment by arfon** [comment]
@whedon generate pdf from branch joss-paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss-paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03349/joss.03349/10.21105.joss.03349.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03349/joss.03349/10.21105.joss.03349.pdf) :page_facing_up: :point_left:

**Comment by Kevin-Mattheus-Moerman** [comment]
@fverdugo, @gaelforget I hope you are well. I'm just checking in to see if you've found your way here and are able to start the review process and that you are able to tick the boxes at the top of this issues. Let me know if you have any questions.

**Comment by fverdugo** [comment]
Hi @Kevin-Mattheus-Moerman! Yes, I can tick the boxes above. I'll try to start the review this week.

**Comment by whedon** [comment]
:wave: @fverdugo, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @gaelforget, please update us on how your review is going (this is an automated reminder).

**Comment by gaelforget** [comment]
> 👋 @gaelforget, please update us on how your review is going (this is an automated reminder).

Hi @Kevin-Mattheus-Moerman 

I will work on the review & check-list this week. Have already been reading through the package docs which look really good now I feel

**Comment by fverdugo** [comment]
Hi @jkrumbiegel @Kevin-Mattheus-Moerman 

In this check box

>  - [ ] Repository: Is the source code for this software available at the repository url?

the link to the repo url points to [AbstractPlotting.jl](https://github.com/JuliaPlots/AbstractPlotting.jl) but I guess (based on previous comment by @jkrumbiegel ) that the repo I should review is [Makie.jl](https://github.com/JuliaPlots/Makie.jl). Another indicator is that  [AbstractPlotting.jl](https://github.com/JuliaPlots/AbstractPlotting.jl) has 90 stars vs [Makie.jl](https://github.com/JuliaPlots/Makie.jl) with 895 stars. So, I believe that the repo for this review should be [Makie.jl](https://github.com/JuliaPlots/Makie.jl)

Can you please confirm or clarify this?

**Comment by jkrumbiegel** [comment]
Yes all code is in the Makie.jl monorepo now, while most of it was in AbstractPlotting.jl when I submitted the paper.

**Comment by fverdugo** [comment]
Hi @jkrumbiegel @Kevin-Mattheus-Moerman  ! find my review below 👇 There are some rather small issues marked with **TODO**. @jkrumbiegel please fix (or clarify) them and I'll update my review accordingly.

## Review

> - [x] Repository: Is the source code for this software available at the repository url?

Yes, source code available at https://github.com/JuliaPlots/Makie.jl

> - [x] License: Does the repository contain a plain-text LICENSE file with the contents of an OSI approved software license?

MIT license file available at https://github.com/JuliaPlots/Makie.jl/blob/master/LICENSE.md

> - [x] Contribution and authorship: Has the submitting author (@jkrumbiegel) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

@SimonDanish (with +2K commits) and @jkrumbiegel (with +1K commits) are the 2 top contributors to the package.

> - [x] Substantial scholarly effort: Does this submission meet the scope eligibility described in the JOSS guidelines

The package has more that 5.8K commits, 97 contributors, and 36 releases. This is clearly more than 3 person months as the minimum requested.

> - [ ] Installation: Does installation proceed as outlined in the documentation?

The package can be easily installed via the Julia package manager with the command `pkg> add GLMakie`. After this I am able to use Makie to generate plots. **TODO** However, the tests launched with `pkg> test GLMakie` fail with error message `ERROR: expected package ReferenceTests [d37af2e0] to be registered`.

> - [ ] Functionality: Have the functional claims of the software been confirmed?

**TODO**: I cannot run the code snipped given in the manuscript to generate Figure 1. The problem is that file `"airportlocations.csv"` is not given in the local folder provided by `assetpath()` nor in here https://github.com/JuliaPlots/Makie.jl/tree/master/assets. I can generate the rest of the subplots without any problem though.

> - [x] Performance: If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

The paper shows how the package can deal with plots with 100M data points.

> - [x] A statement of need: Do the authors clearly state what problems the software is designed to solve and who the target audience is?

The first page of the documentation webpage states "Makie is a high-performance, extendable, and multi-platform plotting ecosystem for the Julia programming language." which clearly shows the goal (high-performance, extendable and multi-platform plotting system) and the target audience (Julia programmers).

 > - [x] Installation instructions: Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.

Installation is straight-forward via the Julia package manager `Pkg`. List of dependencies are defined the Julia-way via a `Project.toml` file.

> - [x] Example usage: Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).

The documentation provides tutorials (e.g., https://makie.juliaplots.org/dev/basic-tutorial.html). Function signatures are also documented using numerous examples. In addition, the README.md file (https://github.com/JuliaPlots/Makie.jl/blob/master/README.md) contains some practical use cases.

> - [x] Functionality documentation: Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?

Yes, see https://makie.juliaplots.org/dev/index.html .

> - [x] Automated tests: Are there automated tests or manual steps described so that the functionality of the software can be verified?

A complete test suite is in place to test the different parts of the library. Test are written via the testing tools provided by Julia and run automatically via Github Actions.

> - [ ] Community guidelines: Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

**TODO** Some extra entries in the README.md file (https://github.com/JuliaPlots/Makie.jl/blob/master/README.md) detailing 1), 2) and 3) would be welcome.

> - [x] Summary: Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?

Yes, sections "Summary" and "Overview" of the manuscript address this point.

> - [x] A statement of need: Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?

Yes. See section "Statement of need" of the manuscript.

> - [ ] State of the field: Do the authors describe how this software compares to other commonly-used packages?

The authors show how this package benefits from the Julia dispatch system (i.e., it has superior extendability), from Julia performance (i.e., it can handle larger data sets), and also show how it allows users to generate interactive plots in a more convenient way (i.e., reactive design). **TODO** Some comment on how Makie.jl compares with Plots.jl (https://github.com/JuliaPlots/Plots.jl), the other major plotting package in Julia, would be welcome. A citation to Plots.jl would be also welcome.

> - [x] Quality of writing: Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?

The paper is well written.

> - [x] References: Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper citation syntax?

Yes.

**Comment by SimonDanisch** [comment]
>The package can be easily installed via the Julia package manager with the command pkg> add GLMakie. After this I am able to use Makie to generate plots. TODO However, the tests launched with pkg> test GLMakie fail with error message ERROR: expected package ReferenceTests [d37af2e0] to be registered.

Good point... Test are supposed to mainly run in the CI, so the local setup has been bitrotting. Should be easy to fix, although a bit annoying with the current monorepo setup...


> Community guidelines: Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

That's a good idea in general.. I hope we can find some time in the next days to add those.

**Comment by jkrumbiegel** [comment]
@fverdugo I have added a citation to Plots.jl and a short paragraph describing the differences.

I have also added contribution guidelines here https://github.com/JuliaPlots/Makie.jl/blob/master/CONTRIBUTING.md

The file "airportlocations.csv" is in the released assets now, you can see an example in the docs here http://makie.juliaplots.org/stable/plotting_functions/scatter.html. The example should therefore work.

The contribution guidelines also mention now that you need to dev `ReferenceTests` from the monorepo to run the tests, although Simon is also trying to improve that process a bit for local testing.

Let me know if there are any other issues!

**Comment by fverdugo** [comment]
Hi @SimonDanisch @jkrumbiegel @Kevin-Mattheus-Moerman,

I went through the check list again and I would recommend the paper for publication. Nice work!

Just a minor comment: there is a typo in the code snippet of the manuscript. Replace `airports_rep` with `a_rep`.

**Comment by SimonDanisch** [comment]
Awesome :) I also fixed the local tests: https://github.com/JuliaPlots/Makie.jl/pull/1190
But we still need to tag a new version for this to work! Will update the contribution docs before we tag...

**Comment by fverdugo** [comment]
and also do not forget the typo in the code snippet in the manuscript !

**Comment by Kevin-Mattheus-Moerman** [comment]
Thanks @fverdugo for your help here!

**Comment by Kevin-Mattheus-Moerman** [comment]
@gaelforget can you update us on where we stand with this review? Thanks again for your help.

**Comment by gaelforget** [comment]
> Functionality: Have the functional claims of the software been confirmed?
> Performance: If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

I am about to check these two items in the list but would suggest reviewing the text or example in the paper in regard to functionality / performance since : 

-  trying to run the 100 million points example crashed my laptop several times (2018 MacBook Pro). New potential user of Makie who might copy and paste the code snippet from the paper might be turned off if this happens to them too (?), especially if they just read that about this _size that many other plotting frameworks would struggle with_ in the paper. Worked for me with `1_000_000` though, so maybe that's better for the code snippet. And maybe mention system requirements for running the example with `100_000_000`. Not sure.

- about _build up a figure piece by piece in GLMakie, explore the data interactively, and then save a publication-quality version with CairoMakie._ it seems unfortunate that when I do this with the code snippet in the paper the two panels on the left hand side turn into blanks. I believe the scatter disappears due to the markersize spec (maybe these are handled differently in Cairo and GL?). The 3D contour issued a warning that _Volume{Tuple{IntervalSets.ClosedInterval{Float32} ... not supported by cairo right now_ (seems fair enough although not ideal since that's the example in the paper).

**Comment by gaelforget** [comment]
> Quality of writing: Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?

Very nice overall so I checked this one in the list too. A few small issues caught my attention though:

- Sentence _To effectively understand ... powerful plotting software._ seems to be missing a verb / seems broken
- `airports_rep` should be `a_rep` , I believe, in the code snippet
- _Makie’s core package is AbstractPlotting.jl_ will need to be revised to reflect the reorganization that took place over the course of the review. I think this is now in the `MakieCore` subfolder. Not sure if the subfolder structure (MakieCore, CairoMakie, etc) is useful to explain in the paper, but may be that would be useful.

**Comment by gaelforget** [comment]
> State of the field: Do the authors describe how this software compares to other commonly-used packages?
> References: Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper citation syntax?

LGTM but maybe for completeness it could be useful to mention GR.jl and Plots.jl

**Comment by gaelforget** [comment]
I see now in https://github.com/JuliaPlots/Makie.jl/tree/joss-paper that the `Plots.jl` and `AbstractPlotting.jl` bits have already been addressed

**Comment by gaelforget** [comment]
> @gaelforget can you update us on where we stand with this review? Thanks again for your help.

Hi @Kevin-Mattheus-Moerman et al

I just went through the checklist and also recommend that the paper be accepted for publication. Great package and great work by the authors!!

ps. a few small issues that should easily be addressed are listed above

**Comment by jkrumbiegel** [comment]
>     * trying to run the 100 million points example crashed my laptop several times

This is a good point, I made the example with a more powerful GPU, but this could indeed be confusing. It's just that 1 million points are not as impressive of a feat, and we wanted to show what can be possible. Maybe the text should qualify that the GPU can be the bottleneck?

> it seems unfortunate that when I do this with the code snippet in the paper the two panels on the left hand side turn into blanks

Also a very good point. 2D vector graphics can never handle 3D data, so maybe the text should make clearer that CairoMakie will handle the 2D subset (with some 3D things working with caveats). So the scatter didn't work for you but the brain did? Maybe the point size is too small for Cairo to build up to an image, while GLMakie handles this case a little differently. Not ideal for an example.

**Comment by Kevin-Mattheus-Moerman** [comment]
Thanks @gaelforget ! :tada:

**Comment by gaelforget** [comment]
> > ```
> > * trying to run the 100 million points example crashed my laptop several times
> > ```
> 
> 
> This is a good point, I made the example with a more powerful GPU, but this could indeed be confusing. It's just that 1 million points are not as impressive of a feat, and we wanted to show what can be possible. Maybe the text should qualify that the GPU can be the bottleneck?

Would make sense to me. Could also maybe add a commented version of this line of code in the snippet with `1_000_000`, right after the one with `100_000_000`, as a hint to the user. Not sure what's best.
 
> > it seems unfortunate that when I do this with the code snippet in the paper the two panels on the left hand side turn into blanks
> 
> 
> Also a very good point. 2D vector graphics can never handle 3D data, so maybe the text should make clearer that CairoMakie will handle the 2D subset (with some 3D things working with caveats). So the scatter didn't work for you but the brain did? Maybe the point size is too small for Cairo to build up to an image, while GLMakie handles this case a little differently. Not ideal for an example.

I believe the scatter worked but unless I remove `markersize = 0.5` I see what's shown below

![paper_example_e](https://user-images.githubusercontent.com/20276764/128201601-e5f61236-4e22-4640-9deb-ab3bc4263829.png)

**Comment by jkrumbiegel** [comment]
@gaelforget Ok I think I have implemented your suggestions now. Fixed the typo, improved the broken sentence, and included more information about the feasibility of the large plot (a comment that a user might need more gpu power for 100 million) and what's possible with CairoMakie vs GLMakie (2D subset with some 3D capabilities).

If you have any other suggestions, let me know!

**Comment by jkrumbiegel** [comment]
Oh and the version here still says ` Version: v.0.17.4` which was AbstractPlotting, we should probably change that to `v.0.15.0` which is the latest Makie.jl

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon generate pdf

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon set v.0.15.0 as version

**Comment by whedon** [comment]
OK. v.0.15.0 is the version.

**Comment by whedon** [comment]
PDF failed to compile for issue #3349 with the following error: 
```
 Can't find any papers to compile :-(

```

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon generate pdf from branch joss-paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss-paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03349/joss.03349/10.21105.joss.03349.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03349/joss.03349/10.21105.joss.03349.pdf) :page_facing_up: :point_left:

**Comment by Kevin-Mattheus-Moerman** [comment]
@gaelforget would you like to check those last changes? Thanks

**Comment by gaelforget** [comment]
> @gaelforget Ok I think I have implemented your suggestions now. Fixed the typo, improved the broken sentence, and included more information about the feasibility of the large plot (a comment that a user might need more gpu power for 100 million) and what's possible with CairoMakie vs GLMakie (2D subset with some 3D capabilities).
> 
> If you have any other suggestions, let me know!

LGTM

**Comment by Kevin-Mattheus-Moerman** [comment]
@jkrumbiegel @SimonDanisch congratulations, given the positive reviews above we will now process this work for acceptance in JOSS. 

- [ ] Since this work is about to be accepted we recommend you proofread the paper yourself one more time and also that you verify that the author names and affiliations are accurate. 
- [ ] Please archive a copy of the software on ZENODO and report back here with the archive DOI. You may process the upload manually or use these automated steps which many find useful: https://guides.github.com/activities/citable-code/. 
- [ ] Please amend (if needed) the ZENODO archive (you can edit information manually) such that meta-data like the title and the author list match that of the JOSS paper. The license also needs to match your software license. FYI you can also ensure, where possible, at this point that ORCID ids are also added to the ZENODO authors. 
- [ ] If the version tag has changed (currently `v.0.15.0`) please let us know.

**Comment by jkrumbiegel** [comment]
Thank you @Kevin-Mattheus-Moerman, I'm happy to hear that! We're still sorting out some issues with the zenodo metadata, but will get back to you once we have done that.

**Comment by Kevin-Mattheus-Moerman** [comment]
@jkrumbiegel @SimonDanisch I hope you are getting on well. Let me know if you have any questions or need help with any of the final steps.

**Comment by jkrumbiegel** [comment]
The problem is that we currently can't edit the zenodo Metadata, as this ability is linked to the account of a former collaborator. I'm still waiting on zenodo's support to transfer ownership.

**Comment by Kevin-Mattheus-Moerman** [comment]
@openjournals/dev  any advice on this? :point_up:

**Comment by arfon** [comment]
> @openjournals/dev any advice on this? ☝️

Is there an archive already in place that we can use here? If so, I propose moving forward with that.

**Comment by Kevin-Mattheus-Moerman** [comment]
@jkrumbiegel see Arfon's comment, can you share the current archive DOI? :point_up: we may be able to move ahead with that for now.

**Comment by jkrumbiegel** [comment]
This is the zenodo archive link for v.0.15.0 https://zenodo.org/record/5106448

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon set 10.5281/zenodo.3735092 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.3735092" target="_blank">10.5281/zenodo.3735092</a> is the archive.

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon set 10.5281/zenodo.5106448 as archive

**Comment by whedon** [comment]
I'm sorry human, I don't understand that. You can see what commands I support by typing:

```
@whedon commands
```

**Comment by arfon** [comment]
@whedon set 10.5281/zenodo.5106448 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.5106448" target="_blank">10.5281/zenodo.5106448</a> is the archive.

**Comment by jkrumbiegel** [comment]
We've checked the paper again and added Simon's orcid, so from my side everything is ready. As far as I understand we can still amend the zenodo metadata later, right? And that wrong metadata won't be reflected in the JOSS paper. I'm not sure what the function of the zenodo doi is in relation to the JOSS paper

**Comment by arfon** [comment]
> I'm not sure what the function of the zenodo doi is in relation to the JOSS paper

The primary function is to have an archive of the software that is outside of GitHub (or any other non-archival venue/platform).

**Comment by Kevin-Mattheus-Moerman** [comment]
@jkrumbiegel just to add to what @arfon said: The Zenodo archive helps create a long term permanent/persistant storage of the software (e.g. since GitHub repositories may be renamed/removed or code can be moved, so these do not share this permanence/stability). Since the archive is there now we are happy to proceed but it is more clearly linked to the JOSS paper if the meta-data (like the title and authors) also matches. Let me know if you have any other questions.

**Comment by Kevin-Mattheus-Moerman** [comment]
@jkrumbiegel I see archived version for `v0.15.2` have also been created https://zenodo.org/record/5269012. Let me know if we should use that archive and upgrade the version tag here again.

**Comment by jkrumbiegel** [comment]
> @jkrumbiegel I see archived version for `v0.15.2` have also been created https://zenodo.org/record/5269012. Let me know if we should use that archive and upgrade the version tag here again. 

I think we can just keep 0.15, there are patch versions coming in all the time. At least that number is round.

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon recommend-accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
PDF failed to compile for issue #3349 with the following error: 
```
 Can't find any papers to compile :-(

```

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon recommend-accept from branch joss-paper

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.5281/zenodo.4907285 is OK

MISSING DOIs

- 10.1109/mcse.2007.55 may be a valid DOI for title: Matplotlib: A 2D graphics environment

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/2558

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/2558, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true from branch joss-paper 
```

**Comment by Kevin-Mattheus-Moerman** [comment]
@jkrumbiegel can you check that potentially missing DOI :point_up:

**Comment by jkrumbiegel** [comment]
The DOI is correct, https://www.doi.org/10.1109/mcse.2007.55 also works. But we need to make one more correction with the example plot, we there's a typo in one subplot title. We'll let you know once that's fixed.

**Comment by SimonDanisch** [comment]
I pushed a fix for the example plot!

**Comment by jkrumbiegel** [comment]
Nice, then I think we're good to go

**Comment by Kevin-Mattheus-Moerman** [comment]
@whedon accept deposit=true from branch joss-paper

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1432978643320516608) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/2563
1. Wait a couple of minutes, then verify that the paper DOI resolves [https://doi.org/10.21105/joss.03349](https://doi.org/10.21105/joss.03349)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by Kevin-Mattheus-Moerman** [comment]
@jkrumbiegel  @arfon I'll leave this issue open to discuss the finalization of the archive editing. Please update us if you hear from Zenodo or once you've edited the information.

**Comment by kthyng** [comment]
Does this still need to be open? Not sure what the outstanding issue was.

**Comment by Kevin-Mattheus-Moerman** [comment]
@kthyng there is an issue with the Zenodo archive, it needs to be edited (e.g. title does not match): https://zenodo.org/record/5106448 but I believe the authors are having permission issues that prevent them from editing it. We are leaving this open until it is resolved.

**Comment by Kevin-Mattheus-Moerman** [comment]
@jkrumbiegel @arfon any updates?

**Comment by jkrumbiegel** [comment]
Zenodo tried to contact the current owner without success, I hope they'll transfer ownership to us soon.

**Comment by arfon** [comment]
> @kthyng there is an issue with the Zenodo archive, it needs to be edited (e.g. title does not match): https://zenodo.org/record/5106448 but I believe the authors are having permission issues that prevent them from editing it. We are leaving this open until it is resolved.

Having the titles match is a strong preference but not essential for JOSS. As such I propose we close this thread unless the authors especially care about this.

**Comment by Kevin-Mattheus-Moerman** [comment]
@arfon okay fine with me. FYI the author set is also different (all contributors are listed on ZENODO).

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.03349/status.svg)](https://doi.org/10.21105/joss.03349)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.03349">
  <img src="https://joss.theoj.org/papers/10.21105/joss.03349/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.03349/status.svg
   :target: https://doi.org/10.21105/joss.03349
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.03349/status.svg)](https://doi.org/10.21105/joss.03349)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss
